components/idp-mgt/org.wso2.carbon.idp.mgt/src/test/java/org/wso2/carbon/idp/mgt/dao/CacheBackedIdPMgtDAOTest.java

```diff
@@ -0,0 +1,1486 @@
+            throws Exception {
+
+        mockStatic(IdentityDatabaseUtil.class);
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDBConnection()).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+        }
+        assertThrows(IdentityProviderManagementException.class, () ->
+                cacheBackedIdPMgtDAO.getIdPsSearch(null, tenantId, tenantDomain, filter));
+    }
+
+    @DataProvider
+    public Object[][] getPaginatedIdPsSearchData() {
+
+        ExpressionNode expressionNode = new ExpressionNode();
+        List<ExpressionNode> expressionNodeList = new ArrayList<>();
+        expressionNodeList.add(expressionNode);
+
+        List<String> attributesList1 = Arrays.asList("id", "name", "description", "isEnabled", "image", "isPrimary");
+        List<String> attributesList2 = Arrays.asList("homeRealmIdentifier", "isFederationHub", "certificate", "alias",
+                "claims", "roles", "federatedAuthenticators", "provisioning");
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, expressionNodeList, 2, 0, "ASC", "NAME", attributesList1, 2},
+                {SAMPLE_TENANT_ID, expressionNodeList, 1, 1, "ASC", "NAME", attributesList2, 1},
+                {SAMPLE_TENANT_ID, expressionNodeList, 2, 0, "DESC", "NAME", attributesList1, 2},
+        };
+    }
+
+    @Test(dataProvider = "getPaginatedIdPsSearchData")
+    public void testGetPaginatedIdPsSearch(int tenantId, List<ExpressionNode> expressionNodes, int limit,
+                                           int offset, String order, String sortBy, List<String> attributes,
+                                           int count) throws Exception {
+
+        mockStatic(IdentityDatabaseUtil.class);
+        try (Connection connection = getConnection(DB_NAME)) {
+            when(IdentityDatabaseUtil.getDBConnection(anyBoolean())).thenReturn(connection);
+            when(IdentityDatabaseUtil.getDataSource()).thenReturn(dataSourceMap.get(DB_NAME));
+            addTestIdps();
+            List<IdentityProvider> idpList = cacheBackedIdPMgtDAO.getPaginatedIdPsSearch(tenantId, expressionNodes,
+                    limit, offset, order, sortBy, attributes);
+            assertEquals(idpList.size(), count);
+        }
+    }
+
+    @DataProvider
+    public Object[][] getPaginatedIdPsSearchExceptionData() {
+
+        ExpressionNode expressionNode = new ExpressionNode();
+        List<ExpressionNode> expressionNodes = new ArrayList<>();
+        expressionNodes.add(expressionNode);
+        List<String> attributes = Arrays.asList("WrongAttribute");
+
+        return new Object[][]{
+                {SAMPLE_TENANT_ID, expressionNodes, 2, 0, "ASC", "NAME", attributes}
```

Comment:
if we don't checking multiple combination then no need to create a data provider.